File: clients/python/src/fibril/internal/topology.py
"""Client-side routing: a cache of queue ownership plus partition selection.

Mirrors the Rust client (``crates/client`` TopologyCache + route_partition). The
cache is connection-free so routing decisions are testable in isolation. The
connection pool that acts on them lives in ``client.py``. Routing is cache-only
and reactive: a cold or standalone cache routes to partition 0, and a misroute is
corrected by a broker redirect rather than a pre-flight lookup.
"""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import Optional

from ..wire import Redirect, TopologyOk, fnv1a


@dataclass
class OwnerEntry:
    endpoint: str
    partitioning_version: int


@dataclass
class PartitioningEntry:
    count: int
    version: int


@dataclass
class Route:
    """The chosen partition and the partitioning version it was decided under."""

    partition: int
    partitioning_version: int


@dataclass
class RoundRobin:
    """Mutable round-robin cursor for keyless publishes."""

    next: int = 0


_QueueKey = tuple[str, int, Optional[str]]
_CountKey = tuple[str, Optional[str]]


class TopologyCache:
    """Cache of queue ownership and partitioning, warmed by topology and redirects."""

    def __init__(self) -> None:
        self.generation = 0
        self.last_refresh_ms = 0.0
        self._by_queue: dict[_QueueKey, OwnerEntry] = {}
        self._counts: dict[_CountKey, PartitioningEntry] = {}

    def lookup(self, topic: str, partition: int, group: Optional[str]) -> Optional[OwnerEntry]:
        return self._by_queue.get((topic, partition, group))

    def partitioning(self, topic: str, group: Optional[str]) -> Optional[PartitioningEntry]:
        return self._counts.get((topic, group))

    def knows_topic(self, topic: str, group: Optional[str]) -> bool:
        """Whether this (topic, group) is declared in the cluster.

        Counts stay populated while an owner is mid-failover, so this remains true
        through a transition and only goes false when the topic is truly absent.
        """
        return self.partitioning(topic, group) is not None

    def is_populated(self) -> bool:
        """Whether the cache holds a real cluster view (at least one declared queue)."""
        return len(self._counts) > 0

    def endpoints(self) -> set[str]:
        """Endpoints that currently own at least one partition."""
        return {entry.endpoint for entry in self._by_queue.values()}

    def replace(self, topology: TopologyOk) -> None:
        """Replace the whole cache from a full topology snapshot."""
        self.generation = topology.generation
        self._by_queue.clear()
        self._counts.clear()
        for queue in topology.queues:
            self._counts[(queue.topic, queue.group)] = PartitioningEntry(
                count=max(queue.partition_count, 1),
                version=queue.partitioning_version,
            )
            # No owner endpoint means the owner node is not in the registry yet
            # (e.g. mid-failover). Keep the count but leave ownership unresolved.
            if queue.owner_endpoint is None:
                continue
            self._by_queue[(queue.topic, queue.partition, queue.group)] = OwnerEntry(
                endpoint=queue.owner_endpoint,
                partitioning_version=queue.partitioning_version,
            )
        # Streams have no group and no per-partition owner entries; they only feed
        # the partitioning cache so a publisher spreads across their partitions.
        for stream in topology.streams:
            self._counts[(stream.topic, None)] = PartitioningEntry(
                count=max(stream.partition_count, 1),
                version=stream.partitioning_version,
            )

    def apply_redirect(self, redirect: Redirect) -> None:
        """Point-update one partition's owner from a redirect."""
        self._by_queue[(redirect.topic, redirect.partition, redirect.group)] = OwnerEntry(
            endpoint=redirect.owner_endpoint,
            partitioning_version=redirect.partitioning_version,
        )

    def invalidate(self, topic: str, partition: int, group: Optional[str]) -> None:
        """Drop one partition's owner, e.g. after it stops being reachable."""
        self._by_queue.pop((topic, partition, group), None)


def route_partition(
    cache: TopologyCache,
    topic: str,
    group: Optional[str],
    key: Optional[bytes],
    round_robin: RoundRobin,
) -> Route:
    """Select the partition for a publish to (topic, group).

    A key routes by ``hash(key) % N``, otherwise round-robin over N. N is the
    authoritative count from the cache. An unknown N (cold cache or standalone)
    routes to partition 0 under version 0, matching a single-partition queue.
    """
    partitioning = cache.partitioning(topic, group)
    version = partitioning.version if partitioning is not None else 0
    count = max(partitioning.count if partitioning is not None else 1, 1)
    if count == 1:
        return Route(partition=0, partitioning_version=version)
    if key is not None:
        index = fnv1a(key) % count
    else:
        index = round_robin.next % count
        round_robin.next += 1
    return Route(partition=index, partitioning_version=version)

File: clients/typescript/src/internal/topology.ts
// Client-side routing state: a cache of queue ownership learned from
// Op.Topology / Op.Redirect, plus the pure partition-selection logic. Mirrors
// the Rust client (crates/client/src/lib.rs TopologyCache + route_partition).
// This module is connection-free on purpose so the routing decisions can be
// tested in isolation; the connection pool that acts on them lives elsewhere.

import type { RedirectMsg, TopologyOkMsg } from "../protocol.js";

const MASK64 = 0xffffffffffffffffn;
const FNV_OFFSET = 0xcbf29ce484222325n;
const FNV_PRIME = 0x100000001b3n;

/**
 * Stable 64-bit FNV-1a hash for partition selection. Must stay byte-for-byte
 * identical to the broker and every other client so a given key always lands on
 * the same partition (per-key ordering).
 */
export function fnv1a(bytes: Uint8Array): bigint {
  let hash = FNV_OFFSET;
  for (const byte of bytes) {
    hash = (hash ^ BigInt(byte)) & MASK64;
    hash = (hash * FNV_PRIME) & MASK64;
  }
  return hash;
}

/** Owner of one queue partition for routing. */
export interface OwnerEntry {
  endpoint: string;
  partitioningVersion: bigint;
}

/** Authoritative partitioning of a logical queue (topic, group). */
export interface PartitioningEntry {
  count: number;
  version: bigint;
}

/**
 * Outcome of routing one publish: the chosen partition and the partitioning
 * version it was decided under. The version is stamped on the wire so the owner
 * can fence a stale routing view.
 */
export interface Route {
  partition: number;
  partitioningVersion: bigint;
}

// Composite Map keys. JSON encoding keeps a null group distinct from an empty
// string group, which a naive string join would not.
const queueKey = (topic: string, partition: number, group: string | null): string =>
  JSON.stringify([topic, partition, group ?? null]);
const countKey = (topic: string, group: string | null): string =>
  JSON.stringify([topic, group ?? null]);

/**
 * Cache of queue ownership and partitioning. Empty (standalone brokers, or a
 * cold client) means "no routing info" and callers fall back to the bootstrap
 * connection.
 */
export class TopologyCache {
  generation = 0n;
  lastRefreshMs = 0;
  #byQueue = new Map<string, OwnerEntry>();
  #counts = new Map<string, PartitioningEntry>();

  lookup(topic: string, partition: number, group: string | null): OwnerEntry | undefined {
    return this.#byQueue.get(queueKey(topic, partition, group));
  }

  /** Authoritative partitioning for (topic, group), if known. */
  partitioning(topic: string, group: string | null): PartitioningEntry | undefined {
    return this.#counts.get(countKey(topic, group));
  }

  /**
   * Whether this (topic, group) is declared in the cluster. Counts stay
   * populated while an owner is mid-failover, so this remains true through a
   * transition and only goes false when the topic is genuinely absent.
   */
  knowsTopic(topic: string, group: string | null): boolean {
    return this.partitioning(topic, group) !== undefined;
  }

  /** Whether the cache holds a real cluster view (at least one declared queue). */
  isPopulated(): boolean {
    return this.#counts.size > 0;
  }

  /** Endpoints that currently own at least one partition. */
  endpoints(): Set<string> {
    const out = new Set<string>();
    for (const entry of this.#byQueue.values()) out.add(entry.endpoint);
    return out;
  }

  /** Replace the whole cache from a full topology snapshot. */
  replace(topology: TopologyOkMsg): void {
    this.generation = topology.generation;
    this.#byQueue.clear();
    this.#counts.clear();
    for (const queue of topology.queues) {
      this.#counts.set(countKey(queue.topic, queue.group), {
        count: Math.max(queue.partition_count, 1),
        version: queue.partitioning_version,
      });
      // No owner endpoints means the owner node is not in the registry yet (e.g.
      // mid-failover). Keep the count but leave ownership unresolved. The wire
      // carries a priority-ordered list; use the first for now (connecting by
      // name resolves service names). Trying the list in order is a later brick.
      const queueOwner = queue.owner_endpoints[0];
      if (queueOwner === undefined) continue;
      this.#byQueue.set(queueKey(queue.topic, queue.partition, queue.group), {
        endpoint: `${queueOwner.host}:${queueOwner.port}`,
        partitioningVersion: queue.partitioning_version,
      });
    }
    // Streams have no group (keyed under group null). They carry per-partition
    // owners just like queues, so a publisher both spreads across partitions and
    // routes each to its owner. A topic is either a stream or a queue, so the
    // group-null cache entries never collide.
    for (const stream of topology.streams ?? []) {
      this.#counts.set(countKey(stream.topic, null), {
        count: Math.max(stream.partition_count, 1),
        version: stream.partitioning_version,
      });
      const streamOwner = stream.owner_endpoints[0];
      if (streamOwner === undefined) continue;
      this.#byQueue.set(queueKey(stream.topic, stream.partition, null), {
        endpoint: `${streamOwner.host}:${streamOwner.port}`,
        partitioningVersion: stream.partitioning_version,
      });
    }
  }

  /** Point-update one partition's owner from a redirect. */
  applyRedirect(redirect: RedirectMsg): void {
    const owner = redirect.owner_endpoints[0];
    if (owner === undefined) return;
    this.#byQueue.set(queueKey(redirect.topic, redirect.partition, redirect.group), {
      endpoint: `${owner.host}:${owner.port}`,
      partitioningVersion: redirect.partitioning_version,
    });
  }

  /** Drop one partition's owner, e.g. after it stops being reachable. */
  invalidate(topic: string, partition: number, group: string | null): void {
    this.#byQueue.delete(queueKey(topic, partition, group));
  }
}

/** Mutable round-robin cursor for keyless publishes. */
export interface RoundRobin {
  next: number;
}

/**
 * Select the partition for a publish to (topic, group):
 * a key routes by hash(key) % N, otherwise round-robin over N. N is the
 * authoritative count from the cache; an unknown N (cold cache or standalone)
 * routes to partition 0 under version 0, which matches a single-partition queue.
 */
export function routePartition(
  cache: TopologyCache,
  topic: string,
  group: string | null,
  key: Uint8Array | null,
  roundRobin: RoundRobin,
): Route {
  const partitioning = cache.partitioning(topic, group);
  const version = partitioning?.version ?? 0n;
  const count = Math.max(partitioning?.count ?? 1, 1);
  if (count === 1) {
    return { partition: 0, partitioningVersion: version };
  }
  let index: number;
  if (key !== null) {
    index = Number(fnv1a(key) % BigInt(count));
  } else {
    const n = roundRobin.next;
    roundRobin.next = n + 1;
    index = n % count;
  }
  return { partition: index, partitioningVersion: version };
}
